apypie/foreman.py

```diff
@@ -0,0 +1,239 @@
+        if self.error:
+            string += f' - {self.error}'
+        return string
+
+    @classmethod
+    def from_exception(cls, exc, msg):
+        """
+        Create a ForemanException from any other Exception
+
+        Especially useful to gather the error message from HTTP responses.
+        """
+        error = None
+        if hasattr(exc, 'response') and exc.response is not None:
+            try:
+                response = exc.response.json()
+                if 'error' in response:
+                    error = response['error']
+                else:
+                    error = response
+            except Exception:  # pylint: disable=broad-except
+                error = exc.response.text
+        return cls(msg=msg, error=error)
+
+
+class ForemanApi(Api):
+    """
+    apypie.Api with default settings and helper functions for Foreman
+    """
+
+    def __init__(self, **kwargs):
+        self.task_timeout = kwargs.pop('task_timeout', 60)
+        self.task_poll = 4
+        kwargs['api_version'] = 2
+        super().__init__(**kwargs)
+
+    def _resource(self, resource):
+        if resource not in self.resources:
+            raise ForemanApiException(msg="The server doesn't know about {0}, is the right plugin installed?".format(resource))
+        return self.resource(resource)
+
+    def _resource_call(self, resource, *args, **kwargs):
+        return self._resource(resource).call(*args, **kwargs)
+
+    def _resource_prepare_params(self, resource, action, params):
+        api_action = self._resource(resource).action(action)
+        return api_action.prepare_params(params)
+
+    def resource_action(self, resource, action, params, options=None, data=None, files=None,  # pylint: disable=too-many-arguments
+                        ignore_task_errors=False):
+        """
+        Perform a generic action on a resource
+
+        Will wait for tasks if the action returns one
+        """
+        resource_payload = self._resource_prepare_params(resource, action, params)
+        if options is None:
+            options = {}
+        try:
+            result = self._resource_call(resource, action, resource_payload, options=options, data=data, files=files)
+            is_foreman_task = isinstance(result, dict) and 'action' in result and 'state' in result and 'started_at' in result
+            if is_foreman_task:
+                result = self.wait_for_task(result, ignore_errors=ignore_task_errors)
+        except Exception as exc:
+            msg = 'Error while performing {0} on {1}: {2}'.format(
+                action, resource, str(exc))
+            raise ForemanApiException.from_exception(exc, msg) from exc
+        return result
+
+    def wait_for_task(self, task, ignore_errors=False):
+        """
+        Wait for a foreman-tasks task, polling it every ``self.task_poll`` seconds.
+
+        Will raise a ForemanApiException when task has not finished in ``self.task_timeout`` seconds.
```

Comment:
And 404 if tasks isn't installed?

Reply:
The implicit hint was that I'd like it to be called out that this needs the plugin to be present.

Reply:
The exception raised is `ForemanApiException: The server doesn't know about foreman_tasks, is the right plugin installed?`

Reply:
I think that's good enough.